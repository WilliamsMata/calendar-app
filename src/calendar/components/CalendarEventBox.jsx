export const CalendarEventBox = ({ event }) => {
  const { title, user } = event;

  return (
    <>
      <strong>{title}</strong>
      <span> - {user.name}</span>
    </>
  );
};
