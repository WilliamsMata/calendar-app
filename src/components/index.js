export * from "./CheckingAuthSpinner";
export * from "./DarkModeBtn";
