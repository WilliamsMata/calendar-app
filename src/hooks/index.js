export * from "./useDarkMode";
export * from "./useLanguage";
export * from "./useUiStore";
