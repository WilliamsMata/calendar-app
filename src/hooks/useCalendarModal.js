import { useEffect, useMemo, useState } from "react";
import { addDays, differenceInSeconds, startOfDay } from "date-fns";
import Swal from "sweetalert2";

import { useCalendarStore, useUiStore } from "./";
import {
  getSweetModalMessageEN,
  getSweetModalMessageES,
  isUserDeviceInSpanish,
} from "../helpers";
import { useDispatch } from "react-redux";
import { toggleSavingEvent } from "../store";

const colors = [
  "#661ae6",
  "#3abff8",
  "#1fb2a5",
  "#fbbd23",
  "#f87272",
  "#d926aa",
];

const { error } = isUserDeviceInSpanish
  ? getSweetModalMessageES()
  : getSweetModalMessageEN();

export const useCalendarModal = () => {
  //* HOOKS
  const dispatch = useDispatch();
  const [formSubmitted, setFormSubmitted] = useState(false);
  const { activeEvent, startSavingEvent, clearActiveEvent } =
    useCalendarStore();
  const { isDateModalOpen, closeDateModal } = useUiStore();

  const [formValues, setFormValues] = useState({
    title: "",
    notes: "",
    start: startOfDay(new Date()),
    end: startOfDay(addDays(new Date(), 1)),
    bgColor: colors[0],
  });

  const titleClass = useMemo(() => {
    if (!formSubmitted) return "";
    return formValues.title.length > 0 ? "" : "input-error";
  }, [formValues.title, formSubmitted]);

  useEffect(() => {
    if (activeEvent !== null) {
      setFormValues({ ...activeEvent });
    }
  }, [activeEvent]);

  //* EVENTS
  const onInputChange = ({ target }) => {
    setFormValues({
      ...formValues,
      [target.name]: target.value,
    });
  };

  // changing: start || end
  const onDateChange = (event, changing) => {
    setFormValues({
      ...formValues,
      [changing]: event,
    });
  };

  const onCloseModal = () => {
    setFormSubmitted(false);
    closeDateModal();
    clearActiveEvent();
  };

  const onSubmit = async (event) => {
    event.preventDefault();
    setFormSubmitted(true);
    dispatch(toggleSavingEvent());

    const timeDifference = differenceInSeconds(
      formValues.end,
      formValues.start
    );

    if (isNaN(timeDifference) || timeDifference <= 0) {
      Swal.fire({
        title: error.title,
        text: error.text,
        icon: "error",
        buttonsStyling: false,
        width: "35rem",
        customClass: {
          confirmButton: "btn btn-outline btn-error",
        },
      });
      document.querySelector("#date-end").classList.add("input-error");
      dispatch(toggleSavingEvent());
      return;
    }

    if (formValues.title.length <= 0) {
      dispatch(toggleSavingEvent());
      return;
    }

    await startSavingEvent(formValues);

    closeDateModal();
    clearActiveEvent();
    setFormSubmitted(false);
  };

  const handleColorClick = (color) => {
    setFormValues({
      ...formValues,
      bgColor: color,
    });
  };

  return {
    //* Properties
    formValues,
    isDateModalOpen,
    titleClass,
    activeEvent,
    colors,

    //* Methods
    onInputChange,
    onDateChange,
    onCloseModal,
    onSubmit,
    handleColorClick,
  };
};
