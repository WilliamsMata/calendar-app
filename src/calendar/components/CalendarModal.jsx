import Modal from "react-modal";
import DatePicker, { registerLocale } from "react-datepicker";
import "react-datepicker/dist/react-datepicker.css";
import es from "date-fns/locale/es";
import { startOfDay } from "date-fns";
import "sweetalert2/dist/sweetalert2.min.css";

import {
  getEnvVariables,
  getFormMessageEN,
  getFormMessageES,
  isUserDeviceInSpanish,
} from "../../helpers";
import { useCalendarModal } from "../../hooks";
import { CloseModalBtn } from "./CloseModalBtn";
import { useSelector } from "react-redux";
import { LoaderSpinnerIcon } from "./LoaderSpinnerIcon";

registerLocale("es", es);

if (getEnvVariables().VITE_MODE !== "test") {
  Modal.setAppElement("#root");
}

const { titleForm, dateStart, dateEnd, eventTitle, note, colorLabel, button } =
  isUserDeviceInSpanish ? getFormMessageES() : getFormMessageEN();

export const CalendarModal = () => {
  const { isSavingEvent } = useSelector((state) => state.calendar);

  const {
    //* Properties
    formValues,
    isDateModalOpen,
    titleClass,
    activeEvent,
    colors,

    //* Methods
    onInputChange,
    onDateChange,
    onCloseModal,
    onSubmit,
    handleColorClick,
  } = useCalendarModal();

  return (
    <Modal
      isOpen={isDateModalOpen}
      onRequestClose={onCloseModal}
      className="relative mx-2 w-full max-w-[30rem] rounded-lg bg-base-100 p-4 outline-none"
      overlayClassName="modal-fondo"
      closeTimeoutMS={200}
    >
      <h1>{activeEvent?._id ? titleForm.editEvent : titleForm.newEvent}</h1>

      <form className="container form-control w-full" onSubmit={onSubmit}>
        <div className="border-b-2 border-base-200 pb-4">
          <div>
            <label className="label">{dateStart.label}</label>
            <DatePicker
              id="date-start"
              minDate={startOfDay(new Date())}
              selected={formValues.start}
              onChange={(event) => onDateChange(event, "start")}
              className="input-bordered input w-full"
              dateFormat="Pp"
              showTimeSelect
              locale={isUserDeviceInSpanish ? "es" : null}
              timeCaption={isUserDeviceInSpanish ? "Hora" : "Time"}
            />
          </div>

          <div>
            <label className="label">{dateEnd.label}</label>
            <DatePicker
              id="date-end"
              minDate={formValues.start}
              selected={formValues.end}
              onChange={(event) => onDateChange(event, "end")}
              className="input-bordered input w-full"
              dateFormat="Pp"
              showTimeSelect
              locale={isUserDeviceInSpanish ? "es" : null}
              timeCaption={isUserDeviceInSpanish ? "Hora" : "Time"}
            />
          </div>
        </div>

        <div className="flex flex-col gap-2">
          <div>
            <label htmlFor="event-title" className="label">
              {eventTitle.label}
            </label>
            <input
              id="event-title"
              type="text"
              className={`input-bordered input w-full ${titleClass}`}
              placeholder={eventTitle.placeholder}
              name="title"
              autoComplete="off"
              value={formValues.title}
              onChange={onInputChange}
            />
            <label
              htmlFor="event-title"
              id="emailHelp"
              className="label-text-alt"
            >
              {eventTitle.altLabel}
            </label>
          </div>

          <div>
            <textarea
              id="note-description"
              type="text"
              className="textarea-bordered textarea w-full resize-none"
              placeholder={note.placeholder}
              rows="3"
              name="notes"
              value={formValues.notes}
              onChange={onInputChange}
            />
            <label
              htmlFor="note-description"
              id="emailHelp"
              className="label-text-alt"
            >
              {note.altLabel}
            </label>
          </div>

          <div>
            <label className="label">{colorLabel}</label>
            <div className="mx-1 flex items-center justify-between gap-4">
              {colors.map((color) => (
                <div
                  key={color}
                  className={`h-8 w-16 cursor-pointer rounded-lg border ${
                    formValues.bgColor === color
                      ? `outline outline-2 outline-offset-2`
                      : ""
                  }`}
                  style={{ backgroundColor: color }}
                  onClick={() => handleColorClick(color)}
                />
              ))}
            </div>
          </div>

          <button
            id="save-btn"
            type="submit"
            className="btn-outline btn-info btn-block btn mt-2 disabled:bg-base-300"
            disabled={isSavingEvent}
          >
            {isSavingEvent ? (
              <LoaderSpinnerIcon />
            ) : (
              <>
                <img
                  src="./icons/save.svg"
                  alt="save logo"
                  className="mr-4 h-6 w-6 transition"
                />
                <span className="tracking-wider">{button}</span>
              </>
            )}
          </button>
        </div>
      </form>

      <CloseModalBtn />
    </Modal>
  );
};
