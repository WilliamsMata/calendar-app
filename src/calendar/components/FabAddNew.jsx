import { addDays, startOfDay } from "date-fns";
import { useCalendarStore, useUiStore } from "../../hooks";

export const FabAddNew = () => {
  const { openDateModal } = useUiStore();
  const { setActiveEvent } = useCalendarStore();

  const onAddNewClick = () => {
    setActiveEvent({
      title: "",
      notes: "",
      start: startOfDay(new Date()),
      end: startOfDay(addDays(new Date(), 1)),
      bgColor: "#661ae6",
      user: {
        _id: "123",
        name: "Williams",
      },
    });
    openDateModal();
  };

  return (
    <button
      id="add-new-btn"
      onClick={onAddNewClick}
      className="btn-info btn-circle btn-lg btn fixed right-4 bottom-4 transition hover:brightness-105 lg:right-8 lg:bottom-6"
    >
      <img
        src="./icons/add.svg"
        alt="add button"
        className="h-8 w-8 transition"
      />
    </button>
  );
};
