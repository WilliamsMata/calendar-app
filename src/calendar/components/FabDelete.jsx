import Swal from "sweetalert2";
import { isUserDeviceInSpanish } from "../../helpers";
import {
  getDeleteSweetModalMessageEN,
  getDeleteSweetModalMessageES,
} from "../../helpers/getMessages";
import { useCalendarStore, useUiStore } from "../../hooks";

const swalButtons = Swal.mixin({
  customClass: {
    confirmButton:
      "btn btn-info min-w-[5rem] mx-2 transition hover:brightness-110",
    cancelButton:
      "btn btn-error min-w-[5rem] mx-2 transition hover:brightness-110",
  },
  buttonsStyling: false,
});

const { question, cancelled } = isUserDeviceInSpanish
  ? getDeleteSweetModalMessageES()
  : getDeleteSweetModalMessageEN();

export const FabDelete = () => {
  const { isDateModalOpen } = useUiStore();
  const { startDeletingEvent, hasEventSelected, clearActiveEvent } =
    useCalendarStore();

  const onDelete = async () => {
    const result = await swalButtons.fire({
      title: question.title,
      text: question.text,
      icon: "warning",
      showCancelButton: true,
      confirmButtonText: question.confirmBtn,
      cancelButtonText: question.cancelBtn,
      reverseButtons: true,
    });

    if (result.isConfirmed) {
      await startDeletingEvent();
    } else if (result.dismiss === Swal.DismissReason.cancel) {
      swalButtons.fire(cancelled.title, cancelled.text, "error");
    }

    clearActiveEvent();
  };

  return (
    <button
      id="delete-btn"
      onClick={onDelete}
      className={`btn-error btn-circle btn-lg btn fixed left-4 bottom-4 transition hover:brightness-105 lg:left-8 lg:bottom-6 ${
        hasEventSelected ? "" : "hidden"
      } ${isDateModalOpen ? "hidden" : ""}`}
    >
      <img
        src="./icons/trash.svg"
        alt="delete button"
        className="h-8 w-8 transition"
      />
    </button>
  );
};
