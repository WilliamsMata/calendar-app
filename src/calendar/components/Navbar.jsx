import { useDarkMode, useLanguage } from "../../hooks";

export const Navbar = () => {
  const { darkMode, switchMode } = useDarkMode();
  const language = useLanguage();

  return (
    <div className="navbar bg-base-100">
      <div className="flex flex-1 items-center pl-2">
        <img src="calendar.svg" alt="calendar image" className="h-8 w-8" />
        <p className="px-2 text-xl normal-case">Williams</p>
      </div>

      <div className="flex gap-4">
        <button className="btn-circle btn" onClick={switchMode}>
          {darkMode ? (
            <img src="sun.svg" alt="light mode" className="h-6 w-6" />
          ) : (
            <img src="moon.svg" alt="dark mode" className="h-6 w-6" />
          )}
        </button>

        <button id="logout-btn" className="btn-outline btn-error btn gap-2">
          <img src="logout.svg" alt="logout btn" className="h-6 w-6" />
          {language === "es" ? "Salir" : "Logout"}
        </button>
      </div>
    </div>
  );
};
