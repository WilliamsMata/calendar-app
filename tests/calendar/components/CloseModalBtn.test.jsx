import { fireEvent, render, screen } from "@testing-library/react";
import { CloseModalBtn } from "../../../src/calendar/components/CloseModalBtn";
import { useCalendarModal } from "../../../src/hooks/useCalendarModal";

jest.mock("../../../src/hooks/useCalendarModal");

describe("Test in <CloseModalBtn />", () => {
  const mockOnCloseModal = jest.fn();

  beforeEach(() => jest.clearAllMocks());

  test("should show the component correctly", () => {
    useCalendarModal.mockReturnValue({ onCloseModal: mockOnCloseModal });
    const { container } = render(<CloseModalBtn />);

    expect(container).toMatchSnapshot();
  });

  test("should call onCloseModal when clicked", () => {
    useCalendarModal.mockReturnValue({ onCloseModal: mockOnCloseModal });
    render(<CloseModalBtn />);

    const btn = screen.getByRole("button");
    fireEvent.click(btn);

    expect(mockOnCloseModal).toHaveBeenCalled();
  });
});
