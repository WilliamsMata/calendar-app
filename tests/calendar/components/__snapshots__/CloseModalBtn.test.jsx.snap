// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in <CloseModalBtn /> should show the component correctly 1`] = `
<div>
  <button
    class="btn-outline btn-circle btn absolute top-5 right-4"
  >
    <svg
      class="h-6 w-6"
      fill="none"
      stroke="currentColor"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M6 18L18 6M6 6l12 12"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
      />
    </svg>
  </button>
</div>
`;
